homeassistant/components/media_player/cambridgeaudio.py

```diff
@@ -0,0 +1,350 @@
+        """Switch to the selected internet radio preset."""
+        self._smdevice.play_preset(self._sources_reverse[source])
+
+    def set_shuffle(self, shuffle):
+        """Enable/disable shuffle play."""
+        if self._audio_source != "media player":
+            self._smdevice.set_shuffle(shuffle)
+
+    def set_volume_level(self, volume):
+        """Set the current volume level. This only works in pre-amp mode."""
+        if self._support_volumecontrol:
+            volume = int(float(self._volume_max) * float(volume))
+            self._smdevice.set_volume(volume)
+
+    def turn_off(self):
+        """Switch the device OFF (default) or into IDLE mode."""
+        _LOGGER.debug("Powering off (%s)", self._power_off_cmd)
+        self._smdevice.power_off(power_state=self._power_off_cmd)
+
+    def turn_on(self):
+        """Turn the device on if it's in IDLE (network standby) mode."""
+        if self._smdevice.get_power_state() == 'idle':
+            self._smdevice.power_on()
+
+    def update(self):
+        """Fetch new state data from the media_player."""
+        dev = self._smdevice
+
+        pwstate = dev.get_power_state()
+        if pwstate is None or pwstate in ['idle', 'off']:
+            self._state = STATE_OFF
+            return
+        self._state = {
+            'PLAYING': STATE_PLAYING,
+            'PAUSED_PLAYBACK': STATE_PAUSED,
+            'STOPPED': STATE_IDLE,
+            'TRANSITIONING': STATE_PLAYING}.get(dev.get_transport_state())
+        self._audio_source = dev.get_audio_source()
+
+        # only need to update volume if the device supports changing it
+        if self._support_volumecontrol:
+            self._muted = dev.get_mute_state()
+            self._volume = dev.get_volume()
+
+        # don't update playback info when nothing is playing
+        if self._state != STATE_PLAYING:
+            return
+
+        if self._audio_source == "media player":
+            self._album_art = dev.get_current_track_info()['albumArtURI']
+            self._artist = dev.get_current_track_info()['artist']
+            self._album = dev.get_current_track_info()['album']
+            self._trackno = dev.get_current_track_info()['origTrackNo']
+            self._title = dev.get_current_track_info()['trackTitle']
+
+            # track position in seconds
+            pos = dev.get_current_track_info()['currentPos']  # hh:mm:ss
+            pos = dt_util.parse_time(pos)
+            pos = timedelta(hours=pos.hour, minutes=pos.minute,
+                            seconds=pos.second).total_seconds()
+            self._position = int(pos)
+
+            # track length in seconds
+            dur = dev.get_current_track_info()['trackLength']
+            dur = dt_util.parse_time(dur)
+            dur = timedelta(hours=dur.hour, minutes=dur.minute,
+                            seconds=dur.second).total_seconds()
+            self._duration = int(dur)
+
+            self._position_updated_at = dt_util.utcnow()
```

Comment:
We shouldn't update the position every time we call update. Instead, if the position that we are at now is the same as the previous position + time since last update, we don't have to update the position and cause a state change.

Reply:
Ok, good point! I changed that now.